### DISCUSSION TITLE
Enable @typescript-eslint/recommended(-type-checked) in create-next-app --typescript

### DISCUSSION THREAD

**Discussion by JoshuaKGoldberg** [vercel/next.js#37151] [Ideas]
### Describe the feature you'd like to request

Right now, Next.js projects created with `create-next-app --typescript` use the TSESLint parser for TypeScript projects -great!- but haven't enabled any of its [preset configurations](https://typescript-eslint.io/docs/linting/configs). Those configurations are recommended for most-to-all TypeScript projects by the TSESLint maintenance team _(including @armano2, @bradzacher, @JamesHenry, and myself; CCing)_. At the very least they'd resolve some of the issues mentioned in #27051 around detecting floating promises.

To verify the list of ESLint rules used in an app newly created with `create-next-app --typescript`, you can run:

```sh
npx eslint --print-config pages/index.tsx > rules.json
```

Is there a particular reason not to include the ~`@typescript-eslint/recommended` and `@typescript-eslint/recommended-requiring-type-checking`~ `@typescript-eslint/recommended-type-checked` _edit: updated for the v6 streamlining of the config_ configs for CNA users by default? 

A couple of notes on why that's likely more desirable now than in the past:

* The recommended configs should no longer contain opinionated rules that wouldn't be good in most projects
* ~If there are still overly chafe-y rules, we're gearing up for a 6.X release and would be happy to move those to the new [strict configuration](https://typescript-eslint.io/docs/linting/configs#strict)~ _Edit: we did this already_ 🙂 

Cheers!

### Describe the solution you'd like

1. Make a `next/typescript` lint config in `eslint-plugin-next` that extends from ~`@typescript-eslint/recommended` and `@typescript-eslint/recommended-requiring-type-checking`~ `@typescript-eslint/recommended-type-checked` _edit: updated for the v6 streamlining of the config_
2. In `create-next-app`, when `--typescript` is provided, change the `.eslintrc.json` file contents to:
    ```json
    {
      "extends": [
        "next/core-web-vitals",
        "next/typescript"
      ]
    }
    ```

Note that this was started in #28181 by @kyliau.

### Describe alternatives you've considered

@kyliau brought up a good point in #28181 that this adds a dependency on the [`@typescript-eslint/eslint-plugin`](https://npmjs.com/package/@typescript-eslint/eslint-plugin) package -which is separate from [`@typescript-eslint/parser`](https://npmjs.com/package/@typescript-eslint/parser)- for JS consumers. An alternative approach could be to have both of those be added to `package.json` and `.eslintrc.js` as described in https://typescript-eslint.io/docs/linting/ only for `--typescript` users.

_Edit July 11th, 2023: renamed from `recommended-requiring-type-checking` to `recommended-type-checked` per [Announcing typescript-eslint v6](https://typescript-eslint.io/blog/announcing-typescript-eslint-v6)_

**Comment by JoshuaKGoldberg** [comment]
I talked a bit with @anthonyshew _(thanks for the advice Anthony!)_ about which benefits of the linting presets are most salient for Next.js & its team. The main points are:

## Embracing the Ecosystem Standard

Extending the recommended rulesets is [how you're strongly encouraged to use typescript-eslint](https://typescript-eslint.io/getting-started). We've been working on getting more of the ecosystem to use them (https://github.com/typescript-eslint/typescript-eslint/issues/6760). For example, [Remix is on the `recommended` config](https://github.com/remix-run/remix/pull/6248) and will soon be on _`recommended-type-checked`_.

A flip side of _not_ using the standard is that Next.js isn't dogfooding/using the rules that many users are. See https://twitter.com/jullerino/status/1676926270309793794 as an example: the [`@typescript-eslint/require-await`](https://typescript-eslint.io/rules/require-await) is a popular and useful rule, but Next.js server actions don't play well with it.

## Fixing Common Issues

The Next.js team has had to deal with a trickle of user reports caused by bad practices that _would_ be caught by the typescript-eslint recommended rules. [`@typescript-eslint/no-floating-promises`](https://typescript-eslint.io/rules/no-floating-promises) is an exemplar of a rule that would help with some of them. "Floating" Promises break stack traces and can cause unexpected program hangs or crashes:
  * @leerob mentioned _"forgetting to await an async function"_ in https://github.com/vercel/next.js/issues/27051#issue-940908168
  * Discussion from this year on how they cause pain: https://github.com/vercel/next.js/pull/44750#discussion_r1083726055
  * Users getting confused by floating Promises: https://www.reddit.com/r/nextjs/comments/13dej9f/i_need_help_with_nextrouter_and_asyncawait/, https://github.com/vercel/next.js/issues/31337 / https://github.com/vercel/next.js/discussions/31278, 

In general, the recommended rules help flag common bugs and misuses of TypeScript - which improves the Next.js dev experience and can help reduce the amount of incoming issue reports.

**Comment by anthonyshew** [comment]
Love it, @JoshuaKGoldberg! I talked to @timneutkens real quick and our idea that it might help with a lower triage burden didn't turn out. However, Tim also did agree this should be done, anyway!

Would you be able to make a PR? Server Actions would be a spot where we would need to pay special attention that I can think of. There may be others. 🤔

**Reply by JoshuaKGoldberg** [reply]
Amazing! Yes, absolutely - I'll send a PR. Thanks!

**Reply by fadomire** [reply]
any chance of enabling @typescript-eslint/recommended-type-checked in next ?

with a proper nextjs config i guess as mentionned here https://github.com/vercel/next.js/discussions/37151#discussioncomment-10432747

it would a great improvement imo

**Comment by rwieruch** [comment]
Glad that someone brought up this topic! For my new Next.js Book I wanted to provide students with standard ESLint rules such as `no-floating-promises` which imo are super important, especially when dealing with serverless environments. However, the setup was quite tedious and I would not want to have beginners to go through it. So yeah, it would be awesome to just activate a eslint-typescript (or have it by default) rule such as `no-floating-promises` 🚀

**Comment by karlhorky** [comment]
Cross-linking this to [the "async functions discouraged in event handler props" topic that has been trending recently](https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/66505#discussioncomment-10411110)

Just want to highlight that enabling the `@typescript-eslint/recommended-type-checked` preset without any further configuration will cause ESLint warnings / errors on async functions passed to event handler props:

(because of the recommended-type-checked configuration of [`@typescript-eslint/no-misused-promises`](https://typescript-eslint.io/rules/no-misused-promises/)):

[`typescript-eslint` Playground](https://typescript-eslint.io/play/#ts=5.5.3&fileType=.tsx&code=JYWwDg9gTgLgBAcigUwIYGMYINwChfIAekscAJsgGaoCuANvJTQHabATNwDCE4HyzGAAoAlHADeuOHBQwaUTgB4ARjRgwOcDlzrB0AawC841AGcAnqzii4hgHxwADAF84Aejt5n%2BIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oFtLlZkiACa1i0Dr0GoMkRNHHRI4MAF8QKoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

`index.tsx`

```ts
import 'react';

export default function Component() {
  return <button onClick={async () => 0} />;
}
```

Error message:

```
Promise-returning function provided to attribute where a void return was expected. eslint (@typescript-eslint/no-misused-promises)
```

![Screenshot 2024-08-23 at 16 51 29](https://github.com/user-attachments/assets/34a39743-b9ed-407b-b9c5-af8ced879753)


Source: https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/66505#discussioncomment-10066351
